### PR TITLE
Fix issue with Breadcrumb not adding excluded with selected values

### PR DESCRIPTION
**Pull request by olamothe**
Add UT for BreadcrumbValuesList

https://coveord.atlassian.net/browse/JSUI-1872

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: test/ui/BreadcrumbValuesListTest.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+      facet.options.numberOfValuesInBreadcrumb = 5;
+      facet.getValueCaption = jasmine.createSpy('getValueCaption').and.callFake((facetValue: IIndexFieldValue) => facetValue.value);
+      facetValues = _.map(FakeResults.createFakeFieldValues('foo', 3), fieldValue => {
+        const created = FacetValue.createFromFieldValue(fieldValue);
+        created.selected = true;
+        return created;
+      });
+    });
+
+    it('should support rendering as a simple string and not an HTML element', () => {
```
Comment (samisayegh) [inline test/ui/BreadcrumbValuesListTest.ts]: How are you testing the 'not an HTML element' part of the test description?

File: test/ui/BreadcrumbValuesListTest.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    it('should display a caption for each selected values', () => {
+      const breadcrumbList = new BreadcrumbValueList(facet, facetValues, BreadcrumbValueElement);
+      const built = breadcrumbList.build();
+      const captions = $$(built).findAll('.coveo-facet-breadcrumb-caption');
+      expect($$(captions[0]).text()).toContain('foo0');
+      expect($$(captions[1]).text()).toContain('foo1');
+      expect($$(captions[2]).text()).toContain('foo2');
+    });
+
+    it('should display a clear button for each selected values', () => {
```
Comment (samisayegh) [inline test/ui/BreadcrumbValuesListTest.ts]: Based of this test, is it possible to add another test that the clear button removes a breadcrumb from the value list?

File: test/ui/BreadcrumbValuesListTest.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    it('should output the title with the facet title option', () => {
+      facet.options.title = 'My facet title';
+      const breadcrumbList = new BreadcrumbValueList(facet, facetValues, BreadcrumbValueElement);
+      const built = breadcrumbList.build();
+      const title = $$(built).find('.coveo-facet-breadcrumb-title');
+      expect($$(title).text()).toContain('My facet title');
+    });
+
+    it('should display a clickable label for each selected values', () => {
```
Comment (samisayegh) [inline test/ui/BreadcrumbValuesListTest.ts]: Is it possible to test the 'clickable' aspect of the label?

File: test/ui/BreadcrumbValuesListTest.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+      facetValues = _.map(FakeResults.createFakeFieldValues('foo', 3), fieldValue => {
+        const created = FacetValue.createFromFieldValue(fieldValue);
+        created.selected = true;
+        return created;
+      });
+    });
+
+    it('should support rendering as a simple string and not an HTML element', () => {
+      facet.options.title = 'My facet title';
+      const breadcrumbList = new BreadcrumbValueList(facet, facetValues, BreadcrumbValueElement);
```
Comment (samisayegh) [inline test/ui/BreadcrumbValuesListTest.ts]: You could combine `BreadcrumbValueList` initialization and the `build()` call into one line since the variable `breadCrumbList` is not referenced otherwise. This is the case for all the tests in the file.

File: test/ui/BreadcrumbValuesListTest.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    });
+
+    it('should display a clear button for each selected values', () => {
+      const breadcrumbList = new BreadcrumbValueList(facet, facetValues, BreadcrumbValueElement);
+      const built = breadcrumbList.build();
+      const clear = $$(built).findAll('.coveo-facet-breadcrumb-clear');
+      expect(clear.length).toBe(3);
+    });
+
+    it('should display excluded values differently', () => {
```
Comment (samisayegh) [inline test/ui/BreadcrumbValuesListTest.ts]: The tests `should display excluded values differently`, `should add the excluded and selected values together when it exceeds the number of values in breadcrumb` and `should show hidden values on clicking show more` seem like they belong together in a describe block with a beforeEach that sets `.excluded` and `.selected`.

File: test/ui/BreadcrumbValuesListTest.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      const breadcrumbList = new BreadcrumbValueList(facet, facetValues, BreadcrumbValueElement);
+      const built = breadcrumbList.build();
+      const andShowMore = $$(built).findAll('.coveo-facet-breadcrumb-multi-count');
+      expect(andShowMore.length).toBe(1);
+      expect($$(andShowMore[0]).text()).toContain('2 more...');
+    });
+
+    it('should show hidden values on clicking show more', () => {
+      facet.options.numberOfValuesInBreadcrumb = 1;
+      facetValues[2].excluded = true;
```
Comment (samisayegh) [inline test/ui/BreadcrumbValuesListTest.ts]: Are there other value combinations of `.excluded()` and `.selected()` that are worth testing? Or the same combination but on two different facet values?

File: test/ui/BreadcrumbValuesListTest.ts (hunk 1 above)
Comment (dms1lva) [reply test/ui/BreadcrumbValuesListTest.ts]: Line 32 there's an assert on the type of the built string.

File: test/ui/BreadcrumbValuesListTest.ts (hunk 3 above)
Comment (olamothe) [reply test/ui/BreadcrumbValuesListTest.ts]: I think i'll change the title of the test. The clickable part would have to be tested in `BreadcrumbValueElement`, since this class is only building a `list` of those, and each single element in in charge of the "clickable" part.

So the test name is indeed misleading.

File: test/ui/BreadcrumbValuesListTest.ts (hunk 2 above)
Comment (olamothe) [reply test/ui/BreadcrumbValuesListTest.ts]: Like above, I think it would be better to test this properly in the `BreadcrumbValueElement` class instead.

File: test/ui/BreadcrumbValuesListTest.ts (hunk 6 above)
Comment (olamothe) [reply test/ui/BreadcrumbValuesListTest.ts]: I could add the test for the edge case where everything is excluded and everything included. 

We're already testing for "some are include and some are excluded", which I think is good enough.

I am not sure what you mean by > same combination but on two different facet values?

Comment (samisayegh) [reply test/ui/BreadcrumbValuesListTest.ts]: I was wondering if there may cases worth testing with facetValue[2] set to `excluded=true` and facetValue[3] set as `selected=false` as opposed to setting both properties on facetValue[2]?

Comment (olamothe) [reply test/ui/BreadcrumbValuesListTest.ts]: Ahhh I see.

Humm testing for edge case make sense and should be done (because those tends to produce bugs), but adding too much test for those kind of small "minor" variation can become kind of annoying/time consuming for not much gain in term of issue prevented, in my opinion.
